webauthn-rp-proxy/src/main.rs

```diff
@@ -0,0 +1,202 @@
+use clap::{App, Arg};
+use serde::{Deserialize, Serialize};
+use serde_json::Result;
+use std::io::{self, Read};
+use uuid::Uuid;
+use webauthn_rs::prelude::*;
+
+// A simple command-line wrapper around webauthn-rs
+
+// There are four commands, each representing a server-side step in
+// the basic webauthn protocol:
+
+// - register-start
+// - register-finish
+// - authenticate-start
+// - authenticate-finish
+
+// The program's --step argument specifies which step from the list
+// above is being invoked.  It reads the corresponding JSON Request
+// struct below from stdin, carries out the webauthn step, and writes
+// the JSON Response struct to stdout if no errors occur.
```

Comment:
The protocol here is strange: the tool will send JSON if everything is fine, but then raw strings if there's an error.

Those errors could be collected up into JSON as well – it's still good to send a non-zero exit code, but it means you can structure the errors a bit better on the other side, as there may be cases where user input (ie: things from the user's authenticator, rather than implementation errors in your app) cause it to be rejected.

Being able to handle that in a structured way on the app's side would be nice.

Essentially you'd serialise a `Result` type, like `Result<RegisterStartResponse, WebauthnError>`. I'm not sure if `serde_json` will serialise a `Result` properly, so you may end up with your own `enum` there instead.

Reply:
You're absolutely right.  I was using the fail-fast approach, relying on the exit code to signal failure, but it's better to collect error information and deliver it to the caller.  Errors are now put into a JSON object that includes the single field `error`, which is a string from the underlying `Err` object.